File: app/server/node/routes/diet.js
const express = require('express');
const router = express.Router();
const db = require('../db');
const config = require('../config');
const fs = require('fs');
const path = require('path');
const logger = require('../logger');

let _recipesCache = null;
let _foodSafetyCache = null;

function _weekToStage(week) {
  if (week <= 0) return 'preparing';
  if (week <= 13) return 'early';
  if (week <= 27) return 'mid';
  if (week <= 42) return 'late';
  return 'nursing';
}

function _convertSafetyToByStage(safety) {
  const mapping = {
    safe:     { preparing: 'safe', early: 'safe', mid: 'safe', late: 'safe', nursing: 'safe' },
    caution:  { preparing: 'safe', early: 'caution', mid: 'caution', late: 'caution', nursing: 'safe' },
    avoid:    { preparing: 'caution', early: 'unsafe', mid: 'caution', late: 'caution', nursing: 'caution' },
    limit:    { preparing: 'limit', early: 'limit', mid: 'limit', late: 'limit', nursing: 'limit' },
  };
  return mapping[safety] || mapping.caution;
}

function loadRecipes() {
  if (_recipesCache) return _recipesCache;
  const filePath = path.join(config.DATA_DIR, 'recipes.json');
  logger.info('diet', `loadRecipes - loading from ${filePath}`);
  if (fs.existsSync(filePath)) {
    const raw = JSON.parse(fs.readFileSync(filePath, 'utf-8'));
    _recipesCache = Array.isArray(raw) ? raw : (raw.recipes || []);
    logger.info('diet', `loadRecipes - loaded ${_recipesCache.length} recipes`);
  } else {
    logger.warn('diet', `loadRecipes - file not found at ${filePath}`);
    _recipesCache = [];
  }
  return _recipesCache;
}

function loadFoodSafety() {
  if (_foodSafetyCache) return _foodSafetyCache;
  const filePath = path.join(config.DATA_DIR, 'food_safety_v3.json');
  logger.info('diet', `loadFoodSafety - loading from ${filePath}`);
  if (fs.existsSync(filePath)) {
    const raw = JSON.parse(fs.readFileSync(filePath, 'utf-8'));
    // Normalize malformed categories data: some categories have {name, icon, items: [...]}
    // while others are flattened as {name, safety, note, safety_by_stage} directly in the array
    const rawCategories = raw.categories || [];
    logger.info('diet', `loadFoodSafety - raw categories count: ${rawCategories.length}`);
    const normalized = [];
    let currentCategory = null;

    for (const entry of rawCategories) {
      // If it has "items" array or has "icon", it's a proper category header
      if (entry.icon || (entry.items && Array.isArray(entry.items))) {
        if (currentCategory) normalized.push(currentCategory);
        currentCategory = {
          name: entry.name,
          icon: entry.icon || '🍽️',
          items: entry.items || [],
        };
      } else if (entry.name) {
        // Flattened item - belongs to current category
        if (!currentCategory) {
          currentCategory = { name: '其他', icon: '🍽️', items: [] };
        }
        currentCategory.items.push({
          name: entry.name,
          safety: entry.safety || 'caution',
          note: entry.note || '',
          safety_by_stage: entry.safety_by_stage || null,
        });
      }
    }
    if (currentCategory) normalized.push(currentCategory);

    // Also ensure each item has safety_by_stage
    for (const cat of normalized) {
      for (const item of cat.items) {
        if (!item.safety_by_stage && item.safety) {
          item.safety_by_stage = _convertSafetyToByStage(item.safety);
        }
      }
    }

    const totalItems = normalized.reduce((sum, cat) => sum + cat.items.length, 0);
    logger.info('diet', `loadFoodSafety - normalized: ${normalized.length} categories, ${totalItems} total items`);
    _foodSafetyCache = { categories: normalized };
  } else {
    logger.warn('diet', `loadFoodSafety - file not found at ${filePath}`);
    _foodSafetyCache = { categories: [] };
  }
  return _foodSafetyCache;
}

router.get('/diet/recipes', (req, res) => {
  try {
    const recipesData = loadRecipes();
    const allParam = req.query.all === 'true' || req.query.all === '1';

    if (allParam) {
      logger.info('diet', `GET /recipes - returning all ${recipesData.length} recipes`);
      return res.json({ code: 0, data: recipesData, message: 'success' });
    }

    const week = parseInt(req.query.week) || 0;
    const stage = _weekToStage(week);
    logger.info('diet', `GET /recipes - week=${week}, stage=${stage}, total recipes=${recipesData.length}`);

    const filteredRecipes = recipesData.filter(recipe => {
      const suitableWeeks = recipe.suitable_weeks || [0, 42];
      if (Array.isArray(suitableWeeks) && suitableWeeks.length >= 2) {
        if (week < suitableWeeks[0] || week > suitableWeeks[1]) return false;
      }
      const suitableStage = recipe.suitable_stage || [];
      if (Array.isArray(suitableStage) && !suitableStage.includes(stage)) return false;
      return true;
    });

    logger.info('diet', `GET /recipes - filtered to ${filteredRecipes.length} recipes for stage=${stage}`);
    res.json({ code: 0, data: filteredRecipes, message: 'success' });
  } catch (error) {
    logger.error('diet', `GET /recipes error: ${error.message}`);
    res.json({ code: 1001, data: null, message: error.message });
  }
});

router.post('/diet/spin', (req, res) => {
  try {
    const body = req.body || {};
    const week = parseInt(req.query.week) || parseInt(body.week) || 0;
    const stage = _weekToStage(week);
    const recipesData = loadRecipes();
    logger.info('diet', `POST /spin - week=${week}, stage=${stage}`);

    const candidates = recipesData.filter(recipe => {
      const suitableWeeks = recipe.suitable_weeks || [0, 42];
      if (Array.isArray(suitableWeeks) && suitableWeeks.length >= 2) {
        if (week < suitableWeeks[0] || week > suitableWeeks[1]) return false;
      }
      const suitableStage = recipe.suitable_stage || [];
      if (Array.isArray(suitableStage) && !suitableStage.includes(stage)) return false;
      return true;
    });

    if (candidates.length === 0) {
      logger.warn('diet', `POST /spin - no matching recipes for week=${week}, stage=${stage}`);
      return res.json({ code: 1001, data: null, message: '没有匹配的食谱' });
    }

    const randomIndex = Math.floor(Math.random() * candidates.length);
    logger.info('diet', `POST /spin - selected recipe "${candidates[randomIndex].name}" from ${candidates.length} candidates`);
    res.json({ code: 0, data: candidates[randomIndex], message: 'success' });
  } catch (error) {
    logger.error('diet', `POST /spin error: ${error.message}`);
    res.json({ code: 1001, data: null, message: error.message });
  }
});

router.get('/diet/food-safety', (req, res) => {
  try {
    const category = req.query.category;
    const foodSafetyData = loadFoodSafety();
    const allCategories = foodSafetyData.categories || [];

    if (category) {
      const cat = allCategories.find(c => c.name === category);
      if (!cat) {
        return res.json({ code: 1001, data: null, message: '未找到该分类' });
      }
      const itemsWithStage = (cat.items || []).map(item => ({
        name: item.name,
        safety_by_stage: item.safety_by_stage || _convertSafetyToByStage(item.safety),
        note: item.note || '',
        image: item.image || null,
      }));
      return res.json({
        code: 0,
        data: [{ name: cat.name, icon: cat.icon, items: itemsWithStage }],
        message: 'success'
      });
    }

    const result = allCategories.map(cat => ({
      name: cat.name,
      icon: cat.icon,
      items: (cat.items || []).map(item => ({
        name: item.name,
        safety_by_stage: item.safety_by_stage || _convertSafetyToByStage(item.safety),
        note: item.note || '',
        image: item.image || null,
      })),
    }));

    res.json({ code: 0, data: result, message: 'success' });
  } catch (error) {
    res.json({ code: 1001, data: null, message: error.message });
  }
});

router.get('/diet/food-safety/search', (req, res) => {
  try {
    const keyword = (req.query.keyword || '').trim();
    if (!keyword) {
      return res.json({ code: 1001, data: null, message: '请输入搜索关键词' });
    }

    const foodSafetyData = loadFoodSafety();
    const results = [];

    (foodSafetyData.categories || []).forEach(cat => {
      (cat.items || []).forEach(item => {
        if ((item.name || '').includes(keyword) || (item.note || '').includes(keyword)) {
          results.push({
            name: item.name,
            safety_by_stage: item.safety_by_stage || _convertSafetyToByStage(item.safety),
            note: item.note || '',
            image: item.image || null,
          });
        }
      });
    });

    res.json({ code: 0, data: results, message: 'success' });
  } catch (error) {
    res.json({ code: 1001, data: null, message: error.message });
  }
});

module.exports = router;

File: app/server/backend/routers/diet.py
"""孕程记 - 饮食 API（食谱推荐 + 食物安全）。"""

import json
import os
import random
from typing import List, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from backend.config import settings
from backend.database import get_db
from backend.schemas.common import ApiResponse
from backend.schemas.diet import RecipeSchema, FoodSafetyCategorySchema, FoodSafetyItemSchema

router = APIRouter()

# 数据缓存
_recipes_cache: Optional[List[dict]] = None
_food_safety_cache: Optional[dict] = None


def _load_recipes() -> List[dict]:
    """加载食谱数据（带缓存）。"""
    global _recipes_cache
    if _recipes_cache is None:
        filepath = os.path.join(settings.DATA_DIR, "recipes.json")
        if os.path.exists(filepath):
            with open(filepath, "r", encoding="utf-8") as f:
                _recipes_cache = json.load(f)
        else:
            _recipes_cache = []
    return _recipes_cache


def _convert_safety_to_by_stage(safety: str) -> dict:
    """将 v3 单值 safety 转换为 safety_by_stage 多阶段对象。

    转换规则（基于孕期营养学常识）:
    - safe:     全阶段安全
    - caution:  备孕/哺乳安全，早中晚谨慎
    - avoid:    备孕谨慎，早期不安全，中晚哺乳谨慎
    - limit:    全阶段限制
    """
    _MAPPING = {
        "safe": {
            "preparing": "safe",
            "early": "safe",
            "mid": "safe",
            "late": "safe",
            "nursing": "safe",
        },
        "caution": {
            "preparing": "safe",
            "early": "caution",
            "mid": "caution",
            "late": "caution",
            "nursing": "safe",
        },
        "avoid": {
            "preparing": "caution",
            "early": "unsafe",
            "mid": "caution",
            "late": "caution",
            "nursing": "caution",
        },
        "limit": {
            "preparing": "limit",
            "early": "limit",
            "mid": "limit",
            "late": "limit",
            "nursing": "limit",
        },
    }
    return _MAPPING.get(safety, _MAPPING["caution"])


def _load_food_safety() -> dict:
    """加载食物安全数据（带缓存 + v3→标准格式转换）。"""
    global _food_safety_cache
    if _food_safety_cache is None:
        filepath = os.path.join(settings.DATA_DIR, "food_safety_v3.json")
        if os.path.exists(filepath):
            with open(filepath, "r", encoding="utf-8") as f:
                raw = json.load(f)
            converted = {"categories": []}
            for cat in raw.get("categories", []):
                new_items = []
                for item in cat.get("items", []):
                    sbs = item.get("safety_by_stage")
                    if not sbs:
                        safety_val = item.get("safety", "caution")
                        sbs = _convert_safety_to_by_stage(safety_val)
                    new_item = {
                        "name": item.get("name", ""),
                        "safety_by_stage": sbs,
                        "note": item.get("note", ""),
                        "image": item.get("image"),
                    }
                    new_items.append(new_item)
                converted["categories"].append({
                    "name": cat.get("name", ""),
                    "icon": cat.get("icon", ""),
                    "items": new_items,
                })
            _food_safety_cache = converted
        else:
            _food_safety_cache = {"categories": []}
    return _food_safety_cache


def _week_to_stage(week: int) -> str:
    """根据孕周判断阶段。

    基于《中国孕前和孕期保健指南》（2023版）：
    - 孕早期：≤13+6周（即≤13周）
    - 孕中期：14-27+6周（即14-27周）
    - 孕晚期：≥28周
    """
    if week < 14:
        return "early"
    elif week < 28:
        return "mid"
    else:
        return "late"


@router.get("/diet/recipes", response_model=ApiResponse[List[RecipeSchema]])
async def get_recipes(
    week: int = Query(default=20, ge=0, le=42, description="孕周"),
    all: bool = Query(default=False, description="是否返回全量食谱（不过滤）"),
):
    """根据孕周获取推荐食谱。all=true 时返回全量。"""
    recipes = _load_recipes()
    if all:
        return ApiResponse(data=recipes)
    stage = _week_to_stage(week)
    matched = []
    for r in recipes:
        suitable_weeks = r.get("suitable_weeks", [0, 42])
        if len(suitable_weeks) >= 2:
            if week < suitable_weeks[0] or week > suitable_weeks[1]:
                continue
        suitable_stage = r.get("suitable_stage", [])
        if stage not in suitable_stage:
            continue
        matched.append(r)
    return ApiResponse(data=matched)


@router.post("/diet/spin", response_model=ApiResponse[RecipeSchema])
async def spin_recipe(
    week: int = Query(default=20, ge=0, le=42, description="孕周"),
):
    """随机推荐一道食谱（转盘功能）。"""
    recipes = _load_recipes()
    stage = _week_to_stage(week)
    matched = []
    for r in recipes:
        suitable_weeks = r.get("suitable_weeks", [0, 42])
        if len(suitable_weeks) >= 2:
            if week < suitable_weeks[0] or week > suitable_weeks[1]:
                continue
        suitable_stage = r.get("suitable_stage", [])
        if stage not in suitable_stage:
            continue
        matched.append(r)
    if not matched:
        return ApiResponse(code=1001, data=None, message="没有匹配的食谱")
    chosen = random.choice(matched)
    return ApiResponse(data=chosen)


@router.get("/diet/food-safety", response_model=ApiResponse[List[FoodSafetyCategorySchema]])
async def get_food_safety(
    category: Optional[str] = Query(None, description="分类名称筛选"),
):
    """获取食物安全数据。"""
    data = _load_food_safety()
    categories = data.get("categories", [])
    if category:
        categories = [c for c in categories if c.get("name") == category]
    return ApiResponse(data=categories)


@router.get("/diet/food-safety/search", response_model=ApiResponse[List[FoodSafetyItemSchema]])
async def search_food(
    keyword: str = Query(..., min_length=1, description="搜索关键词"),
):
    """搜索食物安全信息。"""
    data = _load_food_safety()
    results: List[dict] = []
    for cat in data.get("categories", []):
        for item in cat.get("items", []):
            if keyword in item.get("name", "") or keyword in item.get("note", ""):
                results.append(item)
    return ApiResponse(data=results)
